edgraph/server.go

```diff
@@ -1570,6 +1648,73 @@ func parseRequest(qc *queryContext) error {
 	return validateQuery(qc.dqlRes.Query)
 }
 
+// addQueryIfUnique adds dummy queries in the request for checking whether predicate is unique in the db
+func addQueryIfUnique(qc *queryContext) {
+	ctx := context.WithValue(context.Background(), schema.IsWrite, false)
+	if len(qc.gmuList) == 0 {
+		return
+	}
+
+	qc.uniqueVars = map[string]string{}
+	count := 1
+	buildQuery := ""
+	for gmuIndex, gmu := range qc.gmuList {
+		for rdfIndex, pred := range gmu.Set {
+			predSchema, _ := schema.State().Get(ctx, x.NamespaceAttr(pred.Namespace, pred.Predicate))
+			if !predSchema.Unique {
+				continue
+			}
+			queryVar := fmt.Sprintf("__dgraph_uniquecheck_%v__", strconv.Itoa(count))
+			uniqueVarMapKey := strconv.Itoa(gmuIndex) + strconv.Itoa(rdfIndex)
+
+			// Now, we add a query for a predicate  to check if the value of the
+			// predicate sent in the mutation already exists in the DB.
+			// For example, schema => email: string @unique @index(exact) .
+			// To ensure uniqueness of values of email predicate, we will query for the value
+			// to ensure that we are not adding a duplicate value.
+			// There are following use-cases of mutations which we need to handle:
+			// 1. _:a <email> "example@email.com"  .
+			// 2.  _:a <email> val(queryVariable)  .
+			// In the code below, we construct respective query for the above use-cases viz.
+			//  __dgraph_uniquecheck_1__ as var(func: eq(email,"example@email.com"))
+			//  __dgraph_uniquecheck_1__ as var(func: eq(email,val(queryVariable)))
+			// Each of the above query will check if there is already an existing value.
+			// We can be sure that we may get at most one UID in return.
+			// If the returned UID is different than the UID we have
+			// in the mutation, then we abort the mutation.
+
+			if !strings.HasPrefix(pred.ObjectId, "val(") {
```

Comment:
if and else condition statements are exactly the same except the value in the query. We should just have an if..else on that.